File: 0040-combination-sum-ii/0040-combination-sum-ii.cpp
class Solution {
public:
    void fn(int i, int target, vector<int>& arr, vector<int>& ds, vector<vector<int>>& ans){
        if(target==0){
            ans.push_back(ds);
            return;
        }
        for(int j=i;j<arr.size();j++){
            if(j>i && arr[j-1]==arr[j]) continue;
            if(arr[j]>target) break;
            ds.push_back(arr[j]);
            fn(j+1,target-arr[j],arr,ds,ans);
            ds.pop_back();
        }
    }
    vector<vector<int>> combinationSum2(vector<int>& candidates, int target) {
        sort(candidates.begin(),candidates.end());
        vector<vector<int>> ans;
        vector<int> ds;
        fn(0,target,candidates,ds,ans);
        return ans;
    }
};

File: 0040-combination-sum-ii/0040-combination-sum-ii.py
class Solution:
    def combinationSum2(self, candidates: List[int], target: int) -> List[List[int]]:
        res = []
        candidates.sort()
        
        n = len(candidates)
        
        def dfs(i,cur,total):
            if total == target:
                res.append(cur.copy())
                return
            if total>target or i == n:
                return
            
            cur.append(candidates[i])
            dfs(i+1,cur,total+candidates[i])
            cur.pop()
            
            while i+1<n and candidates[i]==candidates[i+1]:
                i+=1
                
            dfs(i+1,cur,total)
        
        dfs(0,[],0)
        return res